gms/gossiper.cc

```diff
@@ -855,6 +862,46 @@ future<> gossiper::failure_detector_loop() {
     logger.info("failure_detector_loop: Finished main loop");
 }
 
+void gossiper::duplicate_live_endpoints_on_change() {
+    //
+    // Gossiper task runs only on CPU0:
+    //
+    //    - If endpoint_state_map or _live_endpoints have changed - duplicate
+    //      them across all other shards.
+    //    - Reschedule the gossiper only after execution on all nodes is done.
+    //
+    bool live_endpoint_changed = (_live_endpoints != _shadow_live_endpoints);
+    bool unreachable_endpoint_changed = (_unreachable_endpoints != _shadow_unreachable_endpoints);
+
+    if (live_endpoint_changed || unreachable_endpoint_changed) {
+        logger.debug("duplicating live endpoints");
+        if (live_endpoint_changed) {
+            _shadow_live_endpoints = _live_endpoints;
+        }
+
+        if (unreachable_endpoint_changed) {
+            _shadow_unreachable_endpoints = _unreachable_endpoints;
+        }
+
+        container().invoke_on_all([this, live_endpoint_changed, unreachable_endpoint_changed, es = _endpoint_state_map] (gossiper& local_gossiper) {
+            // Don't copy gossiper(CPU0) maps into themselves!
+            if (this_shard_id() != 0) {
+                if (live_endpoint_changed) {
+                    local_gossiper._live_endpoints = _shadow_live_endpoints;
```

Comment:
Hmm, there's a data race there unfortunately.

The code previously assumed that it's not executed concurrently. Therefore it was safe to copy across shards like this.

Unfortunately now it might be executed concurrently: by the gossiper loop and by the API endpoint. This assignment might therefore race with the assigment to `_shadow_live_endpoints` above (and same with `_shardow_unreachable_endpoints`.

We need to fix this somehow. I'm guessing the code was consciously written not to copy these vectors through lambda captures to minimize the number of copies, so we'd like to preserve that. We could maybe place a lock around the copies?